File: client/src/snow/SnowSummary.js
import React, { useState } from 'react'

import { Link, PageSpinner } from '../components/index.js'
import config from '../config.js'
import Snowfalls from './Snowfall.js'
import Condition from './Condition.js'
import Lifts from './Lifts.js'
import Forecasts from './Forecasts.js'

export default function SnowSummary({ summary, isLoading }) {
  let [selected, setSelected] = useState('meadows')
  const selectedSummary = summary?.[selected] || {}
  return (
    <div className="snow-summary-container">
      <section className="snow-summary-conditions">
        {isLoading ? (
          <PageSpinner />
        ) : (
          <>
            <div className="conditions-headers">
              <h1
                onClick={() => setSelected('meadows')}
                className={`${selected === 'meadows' ? 'active' : ''}`}
              >
                Meadows{' '}
                <small>
                  (
                  <Link href={config.meadows.conditionsUrl}>
                    {summary.meadows.lastUpdated}
                  </Link>
                  )
                </small>
              </h1>
              <h1
                onClick={() => setSelected('timberline')}
                className={`${selected === 'timberline' ? 'active' : ''}`}
              >
                Timberline{' '}
                <small>
                  (
                  <Link href={config.timberline.conditionsUrl}>
                    {summary.timberline.lastUpdated}
                  </Link>
                  )
                </small>
              </h1>
              <h1
                onClick={() => setSelected('skiBowl')}
                className={`${selected === 'skiBowl' ? 'active' : ''}`}
              >
                Ski Bowl{' '}
                <small>
                  (
                  <Link href={config.skiBowl.conditionsUrl}>
                    {summary.skiBowl?.lastUpdated || 'Error Updating'}
                  </Link>
                  )
                </small>
              </h1>
            </div>
            <div className="conditions-container">
              <Condition {...selectedSummary.condition} />
              <Snowfalls snowfalls={selectedSummary.snowfalls} />
              <Lifts lifts={selectedSummary.liftStatuses} />
            </div>
          </>
        )}
      </section>
      <section className="snow-summary-noaa">
        {isLoading ? (
          <PageSpinner />
        ) : (
          <>
            <h1>
              NOAA
              <small>
                (<Link href={config[selected].noaaUrl}>go to site</Link>)
              </small>
            </h1>
            <Forecasts forecasts={selectedSummary.forecast} />
          </>
        )}
      </section>
    </div>
  )
}

File: client/src/components/snow_summary.rs
use shared::{ParkStatus, SnowReport};
use yew::prelude::*;

use super::condition::ConditionView;
use super::forecasts::Forecasts;
use super::lifts::LiftsView;
use super::snowfall::SnowfallView;

#[derive(Clone, PartialEq)]
enum ParkTab {
    Meadows,
    Timberline,
    SkiBowl,
}

#[derive(Properties, PartialEq)]
pub struct Props {
    pub report: SnowReport,
}

#[function_component(SnowSummary)]
pub fn snow_summary(props: &Props) -> Html {
    let selected = use_state(|| ParkTab::Meadows);

    let current: &ParkStatus = match *selected {
        ParkTab::Meadows => &props.report.meadows,
        ParkTab::Timberline => &props.report.timberline,
        ParkTab::SkiBowl => &props.report.ski_bowl,
    };

    let make_header = |tab: ParkTab, title: &str| {
        let sel = selected.clone();
        let is_active = *selected == tab;
        let class = if is_active { "active" } else { "" };
        let updated = current
            .status
            .as_ref()
            .and_then(|s| s.updated_on.clone())
            .unwrap_or_else(|| "Unavailable".to_string());
        let tab_clone = tab.clone();
        let onclick = Callback::from(move |_: MouseEvent| sel.set(tab_clone.clone()));
        html! {
            <h1 {onclick} {class}>
                {title}
                <small class="subtitle">{"("}{&updated}{")"}</small>
            </h1>
        }
    };

    html! {
        <div class="snow-summary-container">
            <section class="snow-summary-conditions">
                <div class="conditions-headers">
                    {make_header(ParkTab::Meadows, "Meadows")}
                    {make_header(ParkTab::Timberline, "Timberline")}
                    {make_header(ParkTab::SkiBowl, "Ski Bowl")}
                </div>
                <div class="conditions-container">
                    if let Some(status) = &current.status {
                        if let Some(cond) = &status.condition {
                            <ConditionView condition={cond.clone()} />
                        }
                        <SnowfallView snowfalls={status.snowfalls.clone()} />
                        <LiftsView lifts={status.lifts.clone()} />
                    }
                </div>
            </section>
            <section class="snow-summary-noaa">
                <h1>{"NOAA"}</h1>
                if let Some(status) = &current.status {
                    if let Some(forecasts) = &status.forecast {
                        <Forecasts forecasts={forecasts.clone()} />
                    }
                }
            </section>
        </div>
    }
}
